activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb

```diff
@@ -95,6 +95,26 @@ def test_bad_connection_to_postgres_database
         end
       end
 
+      def test_reconnect_after_bad_connection_on_check_version
+        db_config = ActiveRecord::Base.configurations.configs_for(env_name: "arunit", name: "primary")
+        connection = ActiveRecord::ConnectionAdapters::PostgreSQLAdapter.new(db_config.configuration_hash)
+        connection.connect!
+
+        # mimic a connection that hasn't checked and cached the server version yet i.e. without a raw_connection
+        connection.pool.instance_variable_set(:@server_version, nil)
+        connection.raw_connection.stub(:server_version, 0) do
+          error = assert_raises ActiveRecord::ConnectionFailed do
+            connection.reconnect!
+          end
+          assert_equal "Could not determine PostgreSQL version", error.message
+        end
+
+        # can reconnect after a bad connection
+        assert_nothing_raised do
+          connection.reconnect!
+        end
```

Comment:
Without the patch, if I update the assertion above to expect [this](https://github.com/rails/rails/blob/049d1af38d19968fc8e8648989b0fb8aae2caa71/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L665-L669) `RuntimeError` and the respective message, this assertion fails with:
![Screenshot 2025-03-08 at 12 08 10 am](https://github.com/user-attachments/assets/7b08baf8-5019-42f0-8abf-3ec0c0045e9b)

i.e. it cannot successfully reconnect.